src/solvers/rosenbrock.jl

```diff
@@ -1,93 +1,239 @@
     k = ntuple(n -> similar(prob.u0), Nstages)
-    W = prob.f.jac_prototype
-    linsolve! = alg.linsolve(Val{:init}, W, prob.u0; kwargs...)
-
-    return RosenbrockCache(tab, U, fU, k, W, linsolve!)
+    if !isnothing(prob.f.T_imp!)
+        W = prob.f.T_imp!.jac_prototype
+    else
+        W = nothing
+    end
+    return RosenbrockCache{Nstages, typeof(U), typeof(W)}(U, fU, fU_imp, fU_exp, fU_lim, k, W, ∂Y∂t)
 end
 
+"""
+    step_u!(int, cache::RosenbrockCache{Nstages})
+
+Take one step with the Rosenbrock-method with the given `cache`.
 
-function step_u!(int, cache::RosenbrockCache{Nstages, RT}) where {Nstages, RT}
-    (; m, Γ, A, C) = cache.tableau
+Some choices are being made here. Most of these are empirically motivated and should be
+revisited on different problems.
+- We do not update dtγ across stages
+- We apply DSS to the sum of the explicit and implicit tendency at all the stages but the last
+- We apply DSS to incremented state (ie, after the final stage is applied)
+"""
+function step_u!(int, cache::RosenbrockCache{Nstages}) where {Nstages}
+    (; m, Γ, A, α, C) = int.alg.tableau
     (; u, p, t, dt) = int
-    (; W, U, fU, k, linsolve!) = cache
-    f! = int.sol.prob.f
-    Wfact_t! = int.sol.prob.f.Wfact_t
+    (; W, U, fU, fU_imp, fU_exp, fU_lim, k, ∂Y∂t) = cache
+    T_imp! = int.sol.prob.f.T_imp!
+    T_exp! = int.sol.prob.f.T_exp!
+    T_exp_lim! = int.sol.prob.f.T_exp_T_lim!
+    tgrad! = isnothing(T_imp!) ? nothing : T_imp!.tgrad
+
+    (; post_explicit!, post_implicit!, dss!) = int.sol.prob.f
+
+    has_Wfact_t = false
+
+    # TODO: This is only valid when Γ[i, i] is constant, otherwise we have to
+    # move this in the for loop
+    dtγ = dt * Γ[1, 1]
+
+    if !isnothing(T_imp!)
+        Wfact! = int.sol.prob.f.T_imp!.Wfact
+        Wfact_t! = int.sol.prob.f.T_imp!.Wfact_t
+        if !isnothing(Wfact!) && !isnothing(Wfact_t!)
+            error("Only one between Wfact and Wfact_t can be non-nothing")
+        end
+        if !isnothing(Wfact!)
+            Wfact!(W, u, p, dtγ, t)
+        elseif !isnothing(Wfact_t!)
+            has_Wfact_t = true
+            Wfact_t!(W, u, p, dtγ, t)
+        end
+    end
+
+    if !isnothing(tgrad!)
+        tgrad!(∂Y∂t, u, p, t)
+    end
 
-    # 1) compute jacobian factorization
-    γ = dt * Γ[1, 1]
-    Wfact_t!(W, u, p, γ, t)
     for i in 1:Nstages
+        # Reset tendency
+        fill!(fU, 0)
+
+        αi = sum(α[i, 1:(i - 1)])
+        γi = sum(Γ[i, 1:i])
+
         U .= u
         for j in 1:(i - 1)
             U .+= A[i, j] .* k[j]
         end
-        # TODO: there should be a time modification here (t + c * dt)
-        # if f does depend on time, would need to add tgrad term as well
-        f!(fU, U, p, t)
+
+        if !isnothing(post_implicit!)
+            # NOTE: post_implicit! is a misnomer
+            post_implicit!(U, p, t + αi * dt)
+        end
+
+        if !isnothing(T_imp!)
+            T_imp!(fU_imp, U, p, t + αi * dt)
```

Comment:
The implicit tendency is typically evaluated at a slightly later time than the explicit tendency. Also, the `α` matrix is strictly lower triangular, which corresponds to explicit steps in the context of Butcher tableaus, whereas the diagonals of non-strictly lower triangular matrices correspond to implicit steps. So, I feel like this time should include a contribution from the `Γ` matrix.

Reply:
That's not what I see in the formulation I've been following:
![image](https://github.com/CliMA/ClimaTimeSteppers.jl/assets/9167485/90ffb82d-a541-4584-a329-a00f3b103b1b)

Reply:
Ah, sorry, I forgot this isn't an IMEX method when I made this comment. There's only one tendency here, so it makes sense to evaluate everything at the same time. All good!